crates/ruff/src/rules/flake8_pyi/rules/unused_private_type_declaration.rs

```diff
@@ -0,0 +1,62 @@
+use ruff_diagnostics::{Diagnostic, Violation};
+use ruff_macros::{derive_message_formats, violation};
+use ruff_python_semantic::Binding;
+
+use crate::checkers::ast::Checker;
+
+/// ## What it does
+/// Checks for the presence of unused private `typing.Protocol` definitions.
+///
+/// ## Why is this bad?
+/// A private `typing.Protocol` that is defined but not used is likely a mistake, and should
+/// be removed to avoid confusion.
+///
+/// ## Example
+/// ```python
+/// import typing
+///
+/// class _PrivateProtocol(typing.Protocol):
+///     foo: int
+/// ```
+///
+/// Use instead:
+/// ```python
+/// import typing
+///
+/// class _PrivateProtocol(typing.Protocol):
+///     foo: int
+///
+/// def func(arg: _PrivateProtocol) -> None: ...
+/// ```
+#[violation]
+pub struct UnusedPrivateProtocol {
+    name: String,
+}
+
+impl Violation for UnusedPrivateProtocol {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        let UnusedPrivateProtocol { name } = self;
+        format!("Protocol `{name}` is never used")
```

Comment:
Should we note that it is private in this message? i.e. "Private protocol..."

Reply:
That sounds like a good idea!